calendar-scheduling/src/main/java/com/linagora/calendar/scheduling/AlarmEventScheduler.java

```diff
@@ -0,0 +1,124 @@
+
+import com.linagora.calendar.scheduling.AlarmEventSchedulerConfiguration.Mode;
+import com.linagora.calendar.storage.AlarmEvent;
+import com.linagora.calendar.storage.AlarmEventDAO;
+import com.linagora.calendar.storage.AlarmEventLocker;
+
+import reactor.core.Disposable;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class AlarmEventScheduler implements Startable, Closeable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AlarmEventScheduler.class);
+
+    private final Clock clock;
+    private final AlarmEventDAO alarmEventDAO;
+    private final AlarmEventLocker alarmEventLocker;
+    private final AlarmTriggerService alarmTriggerService;
+    private final AlarmEventSchedulerConfiguration configuration;
+
+    private Disposable loop;
+
+    @Inject
+    @Singleton
+    public AlarmEventScheduler(Clock clock, AlarmEventDAO alarmEventDAO,
+                               AlarmEventLocker alarmEventLockerCandidate,
+                               AlarmTriggerService alarmTriggerService,
+                               AlarmEventSchedulerConfiguration configuration) {
+        this.clock = clock;
+        this.alarmEventDAO = alarmEventDAO;
+        this.alarmTriggerService = alarmTriggerService;
+        this.configuration = configuration;
+
+        if (Mode.SINGLE.equals(configuration.mode())) {
+            alarmEventLocker = AlarmEventLocker.NOOP;
+        } else {
+            alarmEventLocker = alarmEventLockerCandidate;
+        }
+    }
+
+    public void start() {
+        Duration initialDelay = Duration.ofMillis(ThreadLocalRandom.current().nextLong(0,
+            configuration.initialJitterMax().toMillis()));
+
+        LOGGER.info("Starting AlarmEventScheduler: initialDelay={}, pollInterval={}, batchSize={}",
+            initialDelay, configuration.pollInterval(), configuration.batchSize());
+
+        loop = Flux.interval(initialDelay, configuration.pollInterval())
+            .onBackpressureDrop()
+            .concatMap(tick -> pollAndProcess()
+                .onErrorResume(ex -> {
+                    LOGGER.warn("pollAndProcess failed: {}", ex.toString(), ex);
+                    return Mono.empty();
+                }))
+            .doFinally(signal -> LOGGER.info("AlarmEventScheduler terminating, signal={}", signal))
+            .subscribe(count -> {
+                if (count > 0) {
+                    LOGGER.debug("Processed {} alarm(s) this tick", count);
+                }
+            }, ex -> LOGGER.error("AlarmDeliveryWorker loop terminated with error", ex));
+    }
+
+    @Override
+    public void close() {
+        if (loop != null && !loop.isDisposed()) {
+            loop.dispose();
+        }
+    }
+
+    private Mono<Long> pollAndProcess() {
+        return alarmEventDAO.findAlarmsToTrigger(clock.instant())
+            .take(configuration.batchSize())
+            .flatMap(this::processOneAlarm, ReactorUtils.LOW_CONCURRENCY)
```

Comment:
Why taking some alarms. Why not taking all alarms that needs to be triggered?

Note that I'm OK setting a high value to the batch as abusing alarm scheduling is clearly a DOS attack vector but:

 - We can set the batch size to a high value EG 10.000
 - We can warn if that size is exceeded